Mathlib/RingTheory/Length.lean

```diff
@@ -0,0 +1,123 @@
+
+-/
+
+variable (R M : Type*) [CommRing R] [AddCommGroup M] [Module R M]
+
+/-- The length of a module, defined as the krull dimension of its submodule lattice. -/
+noncomputable
+def Module.length : ℕ∞ :=
+  (Order.krullDim (Submodule R M)).unbot (by simp [Order.krullDim_eq_bot_iff])
+
+lemma Module.coe_length :
+    (Module.length R M : WithBot ℕ∞) = Order.krullDim (Submodule R M) :=
+  WithBot.coe_unbot _ _
+
+lemma Module.length_eq_height : Module.length R M = Order.height (⊤ : Submodule R M) := by
+  apply WithBot.coe_injective
+  rw [Module.coe_length, Order.height_top_eq_krullDim]
+
+lemma Module.length_eq_coheight : Module.length R M = Order.coheight (⊥ : Submodule R M) := by
+  apply WithBot.coe_injective
+  rw [Module.coe_length, Order.coheight_bot_eq_krullDim]
+
+variable {R M}
+
+lemma Module.length_eq_zero_iff : Module.length R M = 0 ↔ Subsingleton M := by
+  rw [← WithBot.coe_inj, Module.coe_length, WithBot.coe_zero,
+    Order.krullDim_eq_zero_iff_of_orderTop, Submodule.subsingleton_iff]
+
+@[nontriviality]
+lemma Module.length_eq_zero [Subsingleton M] : Module.length R M = 0 :=
+  Module.length_eq_zero_iff.mpr ‹_›
+
+lemma Module.length_pos_iff : 0 < Module.length R M ↔ Nontrivial M := by
+  rw [pos_iff_ne_zero, ne_eq, Module.length_eq_zero_iff, not_subsingleton_iff_nontrivial]
+
+lemma Module.length_pos [Nontrivial M] : 0 < Module.length R M :=
+  Module.length_pos_iff.mpr ‹_›
+
+lemma Module.length_compositionSeries (s : CompositionSeries (Submodule R M)) (h₁ : s.head = ⊥)
+    (h₂ : s.last = ⊤) : s.length = Module.length R M := by
+  have H := isFiniteLength_of_exists_compositionSeries ⟨s, h₁, h₂⟩
+  have := (isFiniteLength_iff_isNoetherian_isArtinian.mp H).1
+  have := (isFiniteLength_iff_isNoetherian_isArtinian.mp H).2
+  rw [← WithBot.coe_inj, Module.coe_length]
+  apply le_antisymm
+  · let s' := s.map (β := Submodule R M) (s := (· < ·)) ⟨id, fun h ↦ h.1⟩
+    exact (Order.LTSeries.length_le_krullDim s')
+  · rw [Order.krullDim, iSup_le_iff]
+    intro t
+    refine WithBot.coe_le_coe.mpr ?_
+    obtain ⟨t', i, hi, ht₁, ht₂⟩ := t.exists_relSeries_covBy_and_head_eq_bot_and_last_eq_bot
+    have := (s.jordan_holder t' (h₁.trans ht₁.symm) (h₂.trans ht₂.symm)).choose
+    have h : t.length ≤ t'.length := by simpa using Fintype.card_le_of_embedding i
+    have h' : t'.length = s.length := by simpa using (Fintype.card_congr this.symm)
+    simpa using h.trans h'.le
+
+lemma Module.length_ne_top_iff : Module.length R M ≠ ⊤ ↔ IsFiniteLength R M := by
+  constructor
+  · nontriviality M
+    cases finiteDimensionalOrder_or_infiniteDimensionalOrder (Submodule R M)
+    · intro _
+      rw [isFiniteLength_iff_isNoetherian_isArtinian, isNoetherian_iff, isArtinian_iff]
+      exact ⟨Rel.wellFounded_swap_of_finiteDimensional _, Rel.wellFounded_of_finiteDimensional _⟩
+    · rw [ne_eq, ← WithBot.coe_inj, Module.coe_length, WithBot.coe_top, Order.krullDim_eq_top]
+      simp only [not_true_eq_false, IsEmpty.forall_iff]
```

Comment:
slight golf at [this branch](https://github.com/leanprover-community/mathlib4/compare/module_length_golf?expand=1)